lib/invoker.js

```diff
@@ -24,21 +25,53 @@ const invoke = (userFunction) => {
     };
 };
 
-module.exports = (userFunctionUri, port) => {
-    const userFunction = lookUpFunction(userFunctionUri);
-
-    const server = new grpc.Server();
-    server.addService(services.RiffService, {invoke: invoke(userFunction)});
+const executeInit = async (initHook, timeout, server) => {
+    logger('Calling $init hook (if any)');
+    return await guardWithTimeout(initHook, timeout)
+        .then(() => logger('$init hook successfully completed (or skipped if none)'))
+        .catch((err) => {
+            logger(`Error during $init hook execution: ${err.toString()}`);
+            logger('Shutting down now');
+            server.forceShutdown();
+            return Promise.reject('$init error');
+        });
+};
 
+const registerShutdown = (globalScope, destroyHook, timeout, server) => {
     const shutdown = () => {
         logger('Graceful shutdown started');
-        server.tryShutdown(() => logger('Graceful shutdown completed'));
+        logger('Calling $destroy hook (if any)');
+
+        guardWithTimeout(destroyHook, timeout)
+            .then(() => {
+                logger('$destroy hook successfully completed (or skipped if none)');
+                server.tryShutdown(() => logger('Graceful shutdown completed'));
+            })
+            .catch((err) => {
+                logger(`Error during $destroy hook execution: ${err.toString()}`);
+                logger('Shutting down now');
+                server.forceShutdown();
+            });
     };
-    process.on('SIGTERM', shutdown);
-    process.on('SIGINT', shutdown);
+    globalScope.process.once('SIGTERM', shutdown);
+    globalScope.process.once('SIGINT', shutdown);
+};
+
+module.exports = async (globalScope, userFunctionUri, options) => {
+    const server = new grpc.Server();
+
+    const userFunction = lookUpFunction(userFunctionUri);
+
+    await executeInit(userFunction.$init, options.$initTimeoutMs || 10000, server)
+        .then(() => {
+            registerShutdown(globalScope, userFunction.$destroy, options.$destroyTimeoutMs || 10000, server);
```

Comment:
the shutdown should be registered as early as possible and certainly before the init hook is run. If SIGTERM fires during init, it will be lost forever and the process will continue running.

Reply:
good point
my first motivation was to not run the shutdown hook if the init hook failed
I moved the code a bit carelessly here